RFC: Introduce PodmanTestIntegration.PodmanCleanly

mtrmac opened the pull request:
This significantly simplifies the ceromony of running a Podman command in integration tests, from

```go
  session := p.Podman([]string{"stop", id})
  session.WaitWithDefaultTimeout()
  Expect(session).Should(ExitCleanly())
```

to
```go
  p.PodmanCleanly("stop", id)
```

There are >4650 instances of `ExitCleanly()` in the tests, and many could be migrated; this does not do that.

@containers/podman-maintainers RFC. While writing integration tests, I find the ever-present repetitive boilerplate hard to read/maintain, and generally off-putting; OTOH the `PodmanCleanly` call does not show explicitly that any testing is happening. I’m also, short-term, not able to migrate everything, so introducing this would make tests less consistent for now.

It’s also very possible there’s already a much better way to do things, and I just don’t know.

#### Does this PR introduce a user-facing change?

```release-note
None
```

mtrmac commented:
I’m also 100% open to suggestions of a better name of the helper.

rhatdan commented:
I think PodmanExitCleanly() would be a better name.
But this is a good idea.
LGTM

giuseppe reviewed (approved):
LGTM, we can incrementally move new tests to the new function.

I also think `PodmanExitCleanly` is a bit more descriptive

Luap99 commented on test/e2e/common_test.go:
```diff
@@ -484,6 +484,15 @@ func (s *PodmanSessionIntegration) InspectImageJSON() []inspect.ImageData {
+func (p *PodmanTestIntegration) PodmanCleanly(args ...string) *PodmanSessionIntegration {
+	session := p.Podman(args)
+	session.WaitWithDefaultTimeout()
+	Expect(session).Should(ExitCleanly())
```
This should use ExpectWithOffset(1,...), the last line in the stack trace in the log will otherwise all point to this line here which is pretty pointless as I like to get linked to the actual command that failed.

mtrmac replied:
Trying `GinkgoHelper` — that should also affect the failures in `WaitWithDefaultTimeout`.

Luap99 replied:
Ah yes that seems much better than having to deal with correct offset numbers.

Luap99 reviewed (commented):
I also prefer `PodmanExitCleanly` as name. Otherwise I think this is a good idea and I don't see a strict need to manually convert all callers right now.
We can start using it for new code only.

mheon commented:
I like it. LGTM once the comment from @Luap99 is addressed.
I would like to see an existing file migrated (one of the smaller ones) to see what this looks like in practice, but see no need to do this in this PR.

baude commented:
what do people think about the inverse as well?  `PodmanExitFailure` which takes a command and an exit code?

baude commented:
> I like it. LGTM once the comment from @Luap99 is addressed. I would like to see an existing file migrated (one of the smaller ones) to see what this looks like in practice, but see no need to do this in this PR.

If @mtrmac repushes, I can rebase and use in the artifact e2e tests happily.

Luap99 commented:
> what do people think about the inverse as well? `PodmanExitFailure` which takes a command and an exit code?

It would need to take an error message as well (exit code checks are useless without proper error message checks) and then at that point the lines gets so long that it needs to be split on multiple lines anyway so then it is no longer that much shorter. Although I guess getting rid of the WaitWithDefaultTimeout() timeout calls would help a fair bit still

mtrmac commented:
Updated:
- Renamed to `PodmanExitCleanly`
- Updated `attach_test.go` to show what this looks like (notably how the `PodmanExitCleanly` calls look a bit different from raw `Podman` calls when we expect failure or want some extra processing).
- Marked the new utility as a helper.

---

It seems to me that the command tests might be possible to express in some more concise way with a builder-like API:
```go
/* session = */ p.Podman("system", "reset", "-f") // default: run command, default timeout, expect success

/* session = */ p.Command("skopeo", "--version") // other command, default timeout, expect success

/* session = */ p.PodmanWithOptions("other", "situations").Options(
    WithTimeout(other),
    ExpectError(code, string),
    DoNotExecute(), // for truly custom situations
    NoEvents(), // instead of the PodmanBase / Podman split
    …
)
```

_but, also_, we are not in the business of inventing a new integration test DSL. There’s a balance to be struck; the sweet spot is probably being a bit more expressive / capable than just `PodmanExitCleanly` (for non-Podman commands and common error situations?), but not _much_ more expressive, adding dozens of new helpers.

If nothing else, I think that using `...string` instead of having `[]string{}` on every call site is an obvious improvement that could happen in more situations.

Anyway, this is not something I can work on right now. I just wanted to avoid the very obvious boilerplate when writing a few hundred lines of new tests.

Luap99 reviewed (approved):
LGTM
